Remove shrinking from gen-based Prop.forAll

charleso opened the pull request:
Shrinking is only safe when used with Arbitrary. Generators can have a smaller range than the global shrinking, which results in invalid shrunken values.

This fixes issues like #129, #260, #317, #370, #442 and #443.

I suspect this is going to be controversial and I completely understand if it's not accepted. I'm opening this to show how I suggest it should be tackled, but it obviously affects people who are currently relying on this behaviour.

EDIT: Note that Haskell's QuickCheck `forAll` [doesn't shrink in this instance](http://hackage.haskell.org/package/QuickCheck-2.12.4/docs/src/Test.QuickCheck.Property.html#forAll) either

charleso commented:
This breaks binary compatibility (that's a great build check to have). I'm not sure what the suggested approach would be. It can actually be made binary compatible if we keep the `Shrink` implicits but ignore them. That _might_ be safer than having to do a whole new major release. I'm happy to discuss whatever approach people think is more appropriate.

dwijnand commented:
It's probably more acceptable to:

+ keep the implicits in the type signature,
+ don't use them in the implementation,
+ deprecate the methods,
+ Introduce non deprecated methods that doesn't take the Shrink instances,
+ and bump the minor version (x.Y.z).

Or maybe they should be deprecated without changing the behaviour?

charleso commented:
@dwijnand Thanks for the comment/suggestions. Keeping the implicits is probably the safe way to go. I might wait for a conceptual thumbs up from a maintainer though, I'm worried that it may not be something they're interested in.

I'm not as sure about introducing even more methods though. People can already use `forAllNoShrink` now but `forAll` might appear to be the default except for the shrinking issue. I'm also assuming that we can't have the same `forAll` function with and without an implicit, as that would be ambiguous.

Thanks again.

charleso commented:
Sorry, that was a clumsy way of saying that I don’t know whether Rickard or others consider this an issue or not. If people are interested I’m very happy to make whatever changes are required. Thanks again.

dwijnand commented:
Seems I forgot to define the deprecation version in my last change.  I'm logging off, but I can fix it up when I'm back (unless you beat me to it).

charleso commented:
> Seems I forgot to define the deprecation version in my last change. I'm logging off, but I can fix it up when I'm back (unless you beat me to it).

Fixed. There were (funnily enough) a few warnings around calling the deprecated `forAllNoShrink` that I've fixed up now. Hopefully it's all building correctly now. :)

charleso commented:
Another person caught out by the forAll behaviour https://github.com/rickynils/scalacheck/issues/443#issuecomment-443321937

rickynils commented:
@charleso @dwijnand Sorry for my unresponsiveness. I think your approach here makes sense. In one way it makes ScalaCheck less powerful since it removes shrinking that could have been useful in many situations. On the other hand it removes a big source of confusion that a lot of people have stumbled on.

The question is now if we should release a version with unchanged behavior and deprecated methods before actually making this change.

I wonder if we also can remove the hideous `sieve`-thingy from `Gen` when this PR has been merged. It was added as a (poor) attempt at fixing invalid shrinking in some situations.

dwijnand commented:
> The question is now if we should release a version with unchanged behavior and deprecated methods before actually making this change.

If you prefer, I'm happy if you'd like to do things that way.

1.14.1 with deprecated, same behaviour methods, 1.15.0 with this PR?

> I wonder if we also can remove the hideous `sieve`-thingy from `Gen` when this PR has been merged. It was added as a (poor) attempt at fixing invalid shrinking in some situations.

Sounds good.

ashawley commented on src/main/scala/org/scalacheck/Prop.scala:
```diff
@@ -804,81 +802,81 @@ object Prop {
-  /** Universal quantifier for an explicit generator. Shrinks failed arguments
-   *  with the default shrink function for the type */
+  /** Universal quantifier for an explicit generator. */
+  @deprecated("Use the variant without Shrink.  Here only for bincompat.", "1.15.0")
```
How does one still use shrinking? In QuickCheck, the default for `forAll` is no shrinking, but there is still support for shrinking with `forAllShrink`?  From what I see so far, that isn't available here.  This reads like turning off shrinking entirely, when the proposal was changing the default.

charleso replied:
> How does one still use shrinking?

It's still there, but _only_ for the variants of the `forAll` function with the implicit `Arbitrary`. I'm proposing to disable it for the `forAll` functions that take `Gen`.

[def forAll[A1,P] (f: A1 => P)(implicit p: P => Prop, a1: Arbitrary[A1], s1: Shrink[A1]): Prop](https://github.com/rickynils/scalacheck/blob/86fdea6aaebbdf8163324a690f3abdd5fa92593d/src/main/scala/org/scalacheck/Prop.scala#L1012)

charleso replied:
>  In QuickCheck, the default for forAll is no shrinking

Do you mean ScalaCheck? QuickCheck has always only [shrunk with Arbitrary](https://github.com/rickynils/scalacheck/blob/136f7df0b5c3b184ec434dbf6ecc4c5977f91ced/src/main/scala/org/scalacheck/Prop.scala#L900) ([if implemented for a specific instance](http://hackage.haskell.org/package/QuickCheck-2.12.4/docs/src/Test.QuickCheck.Arbitrary.html#shrink)) and [_never_ with Gen](http://hackage.haskell.org/package/QuickCheck-2.12.4/docs/src/Test.QuickCheck.Property.html#forAll), which is what I'm proposing to do in this PR.

ashawley replied:
Maybe I missed the nuance of the change then.

So you're proposing four cases:

1) `forAll` and explicit `Gen`s with shrinking disabled
3) `forAll` with implicit `Arbitrary`s with shrinking preserved
3) `forAllShrink` and explicit `Gen`s with shrinking preserved
4) `forAllNoShrink` with implicit `Arbitrary`s

And still supporting all the possible arities of each.

Right now, ScalaCheck does:

1) `forAll` and explicit `Gen`s with shrinking
3) `forAll` with implicit `Arbitrary`s with shrinking
3) `forAllNoShrink` with explicit `Gen`s with shrinking
4) `forAllNoShrink` with implicit `Arbitrary`s

charleso replied:
> So you're proposing four cases:

Yeah something like that. The `forAllShrink` already exists, but takes a manual shrink function. I'm not touching that. I'm only changing the behaviour of `forAll` with `Gen` which is where the problem is. Everything else stays the same.

I suspect what I've proposed at the moment might have to be tweaked based on @rickynils comment and my own counter-question below (which is totally fine).

ashawley replied:
I sense this will be a problem.  There needs to be a `forAllShrink` with an implicit 
 `Shrink`.  I don't entirely grok the ScalaCheck internals, but I believe the existing `forAllShrink` is not of any help here.

charleso replied:
> I believe the existing forAllShrink is not of any help here.

As a convenient function for taking X number of `Gen`s with implicit shrinking you're quite right. The current function requires the manual shrinking and only a single Gen.

> There needs to be a forAllShrink with an implicit Shrink

On this I'm afraid I don't agree. I'm curious who is building/using manual `Gen` with a restricted subset of the domain (eg. a subset of `Int`, say between 100 and 200) and still want implicit shrinking that might completely go-outside of that domain (eg. shrink to 0)? How do you know what the global/implicit `Shrink` is going to do and whether it aligns with your custom `Gen` range?

I hope this doesn't sound too lazy, but I don't think I would want to add that missing set of functions unless someone _really_ demanded/wanted it. I would try to convince them first it's a bad/dangerous idea. They can still use `Arbitrary` for their own types with their own `Shrink` implementations. :)

ashawley replied:
> I'm curious who is building/using manual `Gen` with a restricted subset of the domain and still want implicit shrinking?

It's a flexibility issue.  Users could be declaring a `Gen` to `forAllShrink` but still defining their own `Shrink` definition in implicit scope, [which is supported in ScalaCheck](https://github.com/rickynils/scalacheck/blob/f31c3ed/doc/UserGuide.md#test-case-minimisation).

charleso replied:
> Users could be declaring a Gen to forAllShrink but still defining their own Shrink definition in implicit scope,

```scala
implicit def ShrinkMyInt: Shrink[Int] = ... // Shrink towards 10 say

//  This will shrink the second generator incorrectly
forAll(Gen.choose(10, 20), Gen.choose(30, 40)) { ... }
```

In the above example you can only provide _one_ implicit for `Shrink[Int]` in scope, but you really need one for _each_ `Gen` otherwise you're just asking for the same kinds of bugs/errors that this PR is trying to fix. I would argue you really shouldn't be doing anything implicit with `Shrink` if you're providing explicit `Gen`. Now perhaps we might want _explicit_ Gen _and_ Shrink together, which is totally fine...

```scala
forAllShrink[A, B](g1: Gen[A], s1: Shrink[A], g2: Gen[B], s2: Shrink[B])(....)
```

At this point I might suggest making a case class with Gen + Shrink to make it easier to pair/keep them together though.

```scala
case class GenShrink[A](gen: Gen[A], shrink: Shrink[A]) 
```

ashawley replied:
There are ways to use different scopes, but supporting generatorso with shrinking having is a usability issue.  Users need the flexibility to study their failing tests by inserting a more limited generator when they are working on an arbitrary property.

For instance, while writing an arbitrary property:

```scala
    forAllShrink { s: String =>
      ...
```

You may want to study the issue with a specific type of generator without giving up shrinking:

```scala
    forAllShrink(Gen.asciiStr) { s: String =>
      ...
```

charleso replied:
> There are ways to use different scopes

Short of passing in each shrinking manually I don't know how you solve the problem of having two generators of the same type (`Char` in this case):

```scala
forAllShrink(Gen.asciiStr, Gen.alphaNumChar) { (a, b) =>
```

There is no shrinker that will work (correctly) for both those cases, so you would need to pass them in manually:

```scala
forAllShrink
  (Gen.asciiStr, Gen.alphaNumChar)
  (Shrink.asciiStr, Shrink.alphaNumChar)
  { (a, b) =>
```

As soon as you pass in one shrink implicit you also have to pass in the extra `A => Pretty` implicit argument as well, which becomes extremely painful. As I said in my previous comment, if you want to keep supporting this you should pass them in explicitly (no implicitly), and I would be tempted to put them together.

```scala
forAllShrink(
    Gen.asciiStr, Shrink.asciiStr
  , Gen.alphaNumChar, Shrink.alphaNumChar
  ) { (a, b) =>
```

charleso replied:
> There are ways to use different scopes

Sorry we probably should have taken this to Gitter at some point. In summary, just to draw a line under this thread.

1. I don't agree that shrinking should _ever_ be "implicit" for Gen (only Arbitrary).
2. I'm more than happy to add whatever extra functionality that @rickynils would like to see, but given that I disagree with Shrinking being implicit for Gen, I'm not likely to add anything along those lines extra unless he asks.

Let me know if you want to continue this conversation on Gitter or some other medium. :)

ashawley replied:
> I don't agree that shrinking should _ever_ be "implicit" for Gen (only Arbitrary).

This seems like a hard line to draw, but we can set it aside.  Instead, given that`forAll` with `Gen` in ScalaCheck presently supports shrinking, it seems it should still be available to users.
And it can be available without interfering with the `forAll` change proposed here, by having it be named `forAllShrink`.

dwijnand replied:
`forAllYoloShrink`? :trollface:

I have no standing, but I guess it would be fine (and maybe a good migration path) to have a forAll that takes an explicit Gen and an implicit Shrink.

ashawley replied:
There are already is one.  We're talking about when arity is greater than one.

dwijnand replied:
I think we're on the same page; by "a forAll" I meant a variant set of forAll's.

ashawley replied:
Ok, but I'm not sure how useful explicit shrinks would be.  I'd say implicits are a feature here.

joroKr21 replied:
FWIW I've been working around this issue for years by using `implicit val noShrink: Shrink[MyType] = Shrink.shrinkAny` (couldn't use `forAllNoShrink` because of scalatest).

I don't think `Gen` with implicit `Shrink`  makes any sense. But I like the idea of combining them. Then the usability issue could be solved by adding a `withShrink(implicit shrink: Shrink[A])` method on `Gen`.

charleso replied:
> Then the usability issue could be solved by adding a withShrink(implicit shrink: Shrink[A]) method on Gen.

I'm curious. I hope you don't mind me asking. What `Gen`s are you constructing/using for your code base, and in particular what is the range of the values? And in what circumstances would you want to summon an `implicit` `Shrink`?

I ask, because I don't really think the implicit part is a good idea (ie `withShrink(shrink: Shrink[A])` would be what I would suggest). But I don't have enough data points to know what people are actually doing with Shrink. I only have my biases. :)

joroKr21 replied:
> I ask, because I don't really think the implicit part is a good idea (ie withShrink(shrink: Shrink[A]) would be what I would suggest). But I don't have enough data points to know what people are actually doing with Shrink. I only have my biases. :)

I tend to agree here, I don't think I would be using the implicitness here very often, but sometimes it might come in handy (say I have my own type with `implicit Shrink[MyType]`) and then I do `Gen.listOf(arbitrary[MyType]).withShrink` then shrinking would be automatically propagated to the list generator, but I could still provide explicit shrinkers for other generators.

woparry replied:
Explicit shrinks (ie. forAllShrink with arity > 1) would be extremely useful for me.

My workflow for working around this issue has always been to write Gen and Shrink instances in parallel, and to wrap them (case class AsciiString(string: String), Gen[AsciiString], Shrink[AsciiString]) if there is ambiguity (because that's less annoying than passing in the Pretty instances).

@charleso 's case class GenShrink[A](gen: Gen[A], shrink: Shrink[A]) looks perfect for this, especially .filter could pretty easily operate on both sides.

charleso replied:
@woparry Thanks for the comment, it's good to hear from someone who is using this.

Just curious, any reason you don't provide an implicit `Arbitrary` to make that `Gen` as the global/implicit version? Is it just that `Gen` works with the global `Shrink`, or something else?

charleso commented:
> The question is now if we should release a version with unchanged behavior and deprecated methods before actually making this change.

I don't particularly have an agenda here, other than to stop people getting confused about the shrinking, I'm super happy to do whatever you think is best for ScalaCheck.

I'm just wondering out-loud what upgrading might look like. If we deprecate users upgrading to 1.14.1 will most likely switch every `forAll` to `forAllNoShrink`. But if they ignore the deprecation warning, and upgrade again to 1.15, the warning goes away.

I guess I'm wondering if we shouldn't re-add the non-shrinking _non-deprecated_ `forAll` back in at all (which is what we're doing in this PR at the moment)? If you're using `Gen` you're always encouraged to use `forAllNoShrinking`? And in 1.15 do we perhaps leave `forAll` with `Gen` deprecated but now the behaviour is not to shrink. Purely visually `forAll` is definitely much shorter/nicer than `forAllNoShrink` but perhaps it makes sense to make the split clear given the previous behaviour and history. I'm not sure though. :(

Again, I'd like to help get this change through in whatever form makes sense. I'm actually pretty indecisive and so I'm happy if you have an opinion on how you want the change(s) implemented and I can do the leg work.

charleso commented:
I've created a PR that just deprecates the `forAll` `Gen` functions as suggested (#444). Let me know if that's a good first step.

charleso commented:
And a second PR #445 which extends that just to disable shrinking but leave the deprecation warning which is slightly different to what this PR is proposing. Sorry for any confusion around having multiple PRs.

ashawley commented:
> I'm just wondering out-loud what upgrading might look like.

An implicit is being deprecated, so the upgrading is the difficult part of this.  Deprecating a method typically will result in its subsequent removal, or a subsequent removal of some signature of the method.  It's not possible to provide a deprecation time period that gives the old version and simultaneously offers a new alternative without the implicits, since the compiler would be confused about conflicting signatures.  And a `deprecatedName` won't work, since presumably nobody is explicitly passing a `Shrink[T]` to `forAll`.  

As a result, when the implicit is removed the user's code will still compile.  If they miss or ignore the deprecation warning, there will be no indication of the behavior change.  Since most test suites succeed when ScalaCheck is upgraded, there won't be any indication that the shrinking policy changed for `forAll(Gen)`.   They would need to be interacting with a failing property test to notice.  So after deprecating, a `migration` annotation should be added when `forAll(Gen)` is changed to mitigate that problem.

It wasn't a poor choice to use an implicit for shrinking in ScalaCheck.  This is a major behavior change in the library, so it's unsurprising this change requires delicate handling. 

The steps I would propose:

1. When `forAll(Gen)` is deprecated, users should be provided a `forAllShrink(Gen)` method (which doesn't exist for all arities up to 8), rather than `forAllNoShrink`.   During deprecation, `forAllShrink(Gen)` will just be an alias to `forAll(Gen)`.
2.  In the release after deprecation, `forAll(Gen)` would become an alias to `forAllNoShrink(Gen)`, and `forAllShrink(Gen)` would remain.  
3. `forAll(Gen)` would have the `deprecated` annotation changed to a `migration`.
4. The `forAllNoShrink(Gen)` is redundant of `forAll(Gen)`, but should be preserved to maintain parity with `forAllNoShrink(Arbitrary)`.

charleso commented:
@ashawley 

I've pushed to #444 with changes to the deprecation warning and the multi-arity `forAllShrink` functions as you've suggested. Please let me know if that looks like what you had in mind?

charleso commented:
@non Thanks again for the detailed response!

I completely agree that a Hedgehog "integrated shrinking" is absolutely 100% the way to go if there's an appetite for it. Even though I'm the author of the Scala Hedgehog library, I strongly believe the wider Scala community would greatly benefit if ScalaCheck gets the same shrinking capability (in the process making Scala Hedgehog mostly redundant).

> Given the new -disableLegacyShrinking flag, do you think there's any other tactical work

I guess I can't help (for the last time I promise) being a broken record and say I still believe disabling shrinking for `forAll` with `Gen` is something we should just do now. I basically see it as a bug fix and bringing it back inline somewhat with the original QuickCheck. It's a source and binary compatible change, and I think it will greatly improve the _default_ interaction with ScalaCheck for _most_ users (as per my [being optimistic comment here](https://github.com/typelevel/scalacheck/pull/522#issuecomment-525925484)). Perhaps it wold make sense to include a `-enableGenShrinking` for users who still really want that ability back (despite the risks).

Anyway, I can see I'm not getting any traction with that argument, and I'm not trying to antagonise anyone over it. I'll close the PR now but I'm happy to re-open or help push something through if there's any subsequent interest. :)

Thanks again for really stepping up to help maintain ScalaCheck, it's already made a huge difference and I'm excited about its future. Good luck!
